### ISSUE TITLE
Complex Union type with generics could not assign return type

### ISSUE THREAD

**Issue by yamachu** [microsoft/TypeScript#43749]
# Bug Report

I don't know which is the correct behavior...
I would appreciate it if you could change the title of the issue.

### 🔎 Search Terms

generic return type union extend

maybe related: https://github.com/microsoft/TypeScript/issues/43425

### 🕗 Version & Regression Information

- This changed between versions 4.3.0-dev.20210315 and 4.3.0-dev.20210323
- I was unable to test this on prior versions (before v3.5.1) because `Omit` was not implemented.

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=next#code/C4TwDgpgBA4glgNwgOwJIBMAq5oF4rICuAtgEYQBOA3ALABQokUAysBXMgOYCGpANhAzYm+ImUq069DsEoAzbgGNoAWRDDoAb3pRdUTohQYAXLENosOSXqgBnNhx79B6U63ZdeAoVfoBfenpGaEwAewAFClDIClAAOUI+PgAeTAAaKABpKAgAD1lkdFsoAGsIEFC5KEwAPih8TSgAbXCoDlLyyuqAXVNWvIKirKgAfgJEvihTTBbuqD9JYKgAQXQsYjAMDFSM7IGUIbKKqtr66oiomPiJnay6gDIoRuANkzsHLnnJIJwWD6cBBozgB5YhwYDJNQaDIAcnsHgBLhhDyeOj08McXhcpgxXEkCygAHpCU99OY3vAkBZoe8EVi3rjOPMfkw4qFkJSjOgAOrggAWmFe6DOq3Wm3Q2zRuncmOc0KlUBhBipGBh9BqVCJJMayq5piISQyjPprlpjgyL3FOP+zKkdEU7PsUFkTvwyQVmBy+QOxRlnjlvwAPlA2RzyTz+YLxeqABQKhDTegASlMMdB4NucP+JuRSagj0axucbz9iJ8TD89TqMe0dBsADpGwg0gqi95TQlxBR62BuBRbIJkMAYwh622XBkAIwABiTLbofiTVCAA)
[worked 4.3.0-dev.20210315](https://www.typescriptlang.org/play?ts=4.3.0-dev.20210315&ssl=1&ssc=1&pln=2&pc=1#code/C4TwDgpgBA4glgNwgOwJIBMAq5oF4rICuAtgEYQBOA3ALABQokUAysBXMgOYCGpANhAzYm+ImUq069DsEoAzbgGNoAWRDDoAb3pRdUTohQYAXLENosOSXqgBnNhx79B6U63ZdeAoVfoBfenpGaEwAewAFClDIClAAOUI+PgAeTAAaKABpKAgAD1lkdFsoAGsIEFC5KEwAPih8TSgAbXCoDlLyyuqAXVNWvIKirKgAfgJEvihTTBbuqD9JYKgAQXQsYjAMDFSM7IGUIbKKqtr66oiomPiJnay6gDIoRuANkzsHLnnJIJwWD6cBBozgB5YhwYDJNQaDIAcnsHgBLhhDyeOj08McXhcpgxXEkCygAHpCU99OY3vAkBZoe8EVi3rjOPMfkw4qFkJSjOgAOrggAWmFe6DOq3Wm3Q2zRuncmOc0KlUBhBipGBh9BqVCJJMayq5piISQyjPprlpjgyL3FOP+zKkdEU7PsUFkTvwyQVmBy+QOxRlnjlvwAPlA2RzyTz+YLxeqABQKhDTegASlMMdB4NucP+JuRSagj0axucbz9iJ8TD89TqMe0dBsADpGwg0gqi95TQlxBR62BuBRbIJkMAYwh622XBkAIwABiTLbofiTVCAA)
[not worked 4.3.0-dev.20210323](https://www.typescriptlang.org/play?ts=4.3.0-dev.20210323#code/C4TwDgpgBA4glgNwgOwJIBMAq5oF4rICuAtgEYQBOA3ALABQokUAysBXMgOYCGpANhAzYm+ImUq069DsEoAzbgGNoAWRDDoAb3pRdUTohQYAXLENosOSXqgBnNhx79B6U63ZdeAoVfoBfenpGaEwAewAFClDIClAAOUI+PgAeTAAaKABpKAgAD1lkdFsoAGsIEFC5KEwAPih8TSgAbXCoDlLyyuqAXVNWvIKirKgAfgJEvihTTBbuqD9JYKgAQXQsYjAMDFSM7IGUIbKKqtr66oiomPiJnay6gDIoRuANkzsHLnnJIJwWD6cBBozgB5YhwYDJNQaDIAcnsHgBLhhDyeOj08McXhcpgxXEkCygAHpCU99OY3vAkBZoe8EVi3rjOPMfkw4qFkJSjOgAOrggAWmFe6DOq3Wm3Q2zRuncmOc0KlUBhBipGBh9BqVCJJMayq5piISQyjPprlpjgyL3FOP+zKkdEU7PsUFkTvwyQVmBy+QOxRlnjlvwAPlA2RzyTz+YLxeqABQKhDTegASlMMdB4NucP+JuRSagj0axucbz9iJ8TD89TqMe0dBsADpGwg0gqi95TQlxBR62BuBRbIJkMAYwh622XBkAIwABiTLbofiTVCAA)

### 💻 Code

```ts
type GivenIdType = number;
type StringableIdType = number;

interface MyType {
    givenId: GivenIdType;
    stringableId: StringableIdType;
}

type ToPropertyNull<T, K extends keyof T> = { [P in keyof T]: P extends K ? null : T[P] };
type AddTmpIdId<T, K extends keyof T> = ToPropertyNull<T, K> & { tmpId: string };

type StringableType = Omit<MyType, 'stringableId'> & {
    stringableId: string;
}; // { givenId: GivenIdType, stringableId: string }
type NonGivenIdWithTmpId = AddTmpIdId<
    StringableType,
    'givenId'
>; // { givenId: null, stringableId: string, tmpId: string }

const test = <
    T extends StringableType | NonGivenIdWithTmpId
>(
    v: T
): (Omit<T, 'stringableId'>) & { stringableId: StringableIdType } => ({
    ...v,
    stringableId: Number.parseInt(v.stringableId, 10),
});
```

### 🙁 Actual behavior

cannot assign function argument to return type

> Type '{ stringableId: number; givenId: number; } | { stringableId: number; givenId: null; tmpId: string; }' is not assignable to type 'Omit<T, "stringableId"> & { stringableId: number; }'.

### 🙂 Expected behavior

Union type function arguments given by generics can be used as part of the return type

**Comment by Kiikurage** [comment]
[Simpler reproducible code](https://www.typescriptlang.org/play?ts=4.3.0-dev.20210323&ssl=1&ssc=1&pln=20&pc=2#code/PTAEHUEsBcAsHsCu1QHMCmA7dAnSBjAZwC5QB5AaQFgAoAM0U32knk1DoEYAKAN1IDeoAA6lC0PJlSgAvqAA+oIaNCZEAWwBGuWQEpBIsRMhTZCpYdUbtOMwNqhHoHOmiIc7IQDofvWQG5aGVpaEHJMABsAT1BGVkxSSloGJhY2DgAmAB4AFVB0AA9oLAATQgsVcUlpOUVlUjUtHRkAPj5SHP1QPPsaJ2dXd3ZeIJCaMLJImMJhFwBDEtB4YVw56HgcROp6RmZ4jgBmXPyi0vL60CqTGrb+bq6ehycXNw8LHy8-YJpv0LAAIXgcFimH2c0wixm80Wy1W602oAAcgBxZK7NLsOgAFmOhWKEPOliuplqFQa1matw6DyUT0cYQAojgcBsOlEVqAAOQXYmofxmOqWRo2fkyTmgSDlTBA0BzQiESCoTBzTQRdCgdYa9nqzk5TleOkDV6eUAfL5BIA)

```ts
// Without generics: OK
function f1(v: { p: string } | { p: number }): { p: string } | { p: number } {
    return { ...v };
}

// Only union: OK
function f2<T extends { p: string } | { p: number }>(v: T): T {
    return v
}

// Only spread operator: OK
function f3<T extends { p: string }>(v: T): T {
    return { ...v }
}

// Both union and spread operator: NG
function f4<T extends { p: string } | { p: number }>(v: T): T {
    // Error: Type '{ p: string; } | { p: number; }' is not assignable to type 'T'.
    return { ...v }
}
```

**Comment by markusjohnsson** [comment]
I found this workaround:

```ts
function workaround<T>(t: T): asserts t is T { }

// Both union and spread operator
function f4<T extends { p: string } | { p: number }>(v: T): T {

    workaround<T>(v);

    // No more error
    return { ...v }
}
```

**Comment by markusjohnsson** [comment]
Looks like this is fixed in nightly build (by #44081  ?)

**Comment by yamachu** [comment]
Yes!

https://www.typescriptlang.org/play?ts=4.4.0-dev.20210517#code/C4TwDgpgBA4glgNwgOwJIBMAq5oF4rICuAtgEYQBOA3ALABQokUAysBXMgOYCGpANhAzYm+ImUq069DsEoAzbgGNoAWRDDoAb3pRdUTohQYAXLENosOSXqgBnNhx79B6U63ZdeAoVfoBfenpGaEwAewAFClDIClAAOUI+PgAeTAAaKABpKAgAD1lkdFsoAGsIEFC5KEwAPih8TSgAbXCoDlLyyuqAXVNWvIKirKgAfgJEvihTTBbuqD9JYKgAQXQsYjAMDFSM7IGUIbKKqtr66oiomPiJnay6gDIoRuANkzsHLnnJIJwWD6cBBozgB5YhwYDJNQaDIAcnsHgBLhhDyeOj08McXhcpgxXEkCygAHpCU99OY3vAkBZoe8EVi3rjOPMfkw4qFkJSjOgAOrggAWmFe6DOq3Wm3Q2zRuncmOc0KlUBhBipGBh9BqVCJJMayq5piISQyjPprlpjgyL3FOP+zKkdEU7PsUFkTvwyQVmBy+QOxRlnjlvwAPlA2RzyTz+YLxeqABQKhDTegASlMMdB4NucP+JuRSagj0axucbz9iJ8TD89Tq2joNgdyCdhezxfQGQAdB2KBAnZX8AhrHou8BCBRkKjazY9B22137GkFTYi95TQlxBQ22BuBRbIJkMAYwg20uXBkAIwABiT84nugCdAWQA

After applying the following change, it works.
```diff
const test = <
    T extends StringableType | NonGivenIdWithTmpId
>(
    v: T
): (Omit<T, 'stringableId'>) & { stringableId: StringableIdType } => {
    // omit
    const { stringableId, ...rest } = v;
    return {
        ...rest,
        // bind
        stringableId: Number.parseInt(v.stringableId, 10),
    }
};
```

Thank you for noticing!

**Comment by ahejlsberg** [comment]
Fixed by #44081. Closing.
